src/python/pants/source/wrapped_globs.py

```diff
@@ -104,7 +104,11 @@ def __repr__(self):
     return 'EagerFilesetWithSpec(rel_root={!r}, files={!r})'.format(self.rel_root, self.files)
 
   def matches(self, path_from_buildroot):
-    path_relative_to_rel_root = os.path.relpath(path_from_buildroot, self.rel_root)
+    try:
+      path_relative_to_rel_root = fast_relpath(path_from_buildroot, self.rel_root)
+    except ValueError:
+      # N.B. `fast_relpath` raises a `ValueError` if prefixes aren't aligned.
```

Comment:
What would this case mean here? That a target owns a source... outside the buildroot? It's not clear that we would want to hide that.

It shouldn't be possible at all given the restrictions we place on symlinks in the engine, so would be good to know why/if you saw it.

Reply:
no, fwict just that `EagerFilesetWithSpec.matches()` is a miss because of unaligned prefixes.

without this, I was seeing tracebacks like this in `--changed list`:

```
Exception caught: (<type 'exceptions.ValueError'>)
  File "/Users/kwilson/dev/pants/src/python/pants/bin/pants_loader.py", line 73, in <module>
    main()
  File "/Users/kwilson/dev/pants/src/python/pants/bin/pants_loader.py", line 69, in main
    PantsLoader.run()
  File "/Users/kwilson/dev/pants/src/python/pants/bin/pants_loader.py", line 65, in run
    cls.load_and_execute(entrypoint)
  File "/Users/kwilson/dev/pants/src/python/pants/bin/pants_loader.py", line 58, in load_and_execute
    entrypoint_main()
  File "/Users/kwilson/dev/pants/src/python/pants/bin/pants_exe.py", line 39, in main
    PantsRunner(exiter, start_time=start_time).run()
  File "/Users/kwilson/dev/pants/src/python/pants/bin/pants_runner.py", line 53, in run
    return runner.run()
  File "/Users/kwilson/dev/pants/src/python/pants/bin/local_pants_runner.py", line 46, in run
    self._run()
  File "/Users/kwilson/dev/pants/src/python/pants/bin/local_pants_runner.py", line 91, in _run
    self._exiter).setup()
  File "/Users/kwilson/dev/pants/src/python/pants/bin/goal_runner.py", line 188, in setup
    goals, context = self._setup_context()
  File "/Users/kwilson/dev/pants/src/python/pants/bin/goal_runner.py", line 159, in _setup_context
    self._global_options.subproject_roots,
  File "/Users/kwilson/dev/pants/src/python/pants/bin/goal_runner.py", line 109, in _init_graph
    change_calculator=graph_helper.change_calculator)
  File "/Users/kwilson/dev/pants/src/python/pants/init/target_roots.py", line 67, in create
    changed_addresses = change_calculator.changed_target_addresses(changed_request)
  File "/Users/kwilson/dev/pants/src/python/pants/engine/legacy/change_calculator.py", line 133, in changed_target_addresses
    return list(self.iter_changed_target_addresses(changed_request))
  File "/Users/kwilson/dev/pants/src/python/pants/engine/legacy/change_calculator.py", line 111, in iter_changed_target_addresses
    in self._mapper.iter_target_addresses_for_sources(changed_files))
  File "/Users/kwilson/dev/pants/src/python/pants/engine/legacy/change_calculator.py", line 109, in <genexpr>
    changed_addresses = set(address
  File "/Users/kwilson/dev/pants/src/python/pants/engine/legacy/source_mapper.py", line 90, in iter_target_addresses_for_sources
    any(self._owns_source(source, hydrated_target) for source in sources_set)):
  File "/Users/kwilson/dev/pants/src/python/pants/engine/legacy/source_mapper.py", line 90, in <genexpr>
    any(self._owns_source(source, hydrated_target) for source in sources_set)):
  File "/Users/kwilson/dev/pants/src/python/pants/engine/legacy/source_mapper.py", line 69, in _owns_source
    if target_sources and self._match_source(source, target_sources):
  File "/Users/kwilson/dev/pants/src/python/pants/engine/legacy/source_mapper.py", line 54, in _match_source
    return fileset.matches(source) or matches_filespec(source, fileset.filespec)
  File "/Users/kwilson/dev/pants/src/python/pants/source/wrapped_globs.py", line 108, in matches
    path_relative_to_rel_root = fast_relpath(path_from_buildroot, self.rel_root)
  File "/Users/kwilson/dev/pants/src/python/pants/util/dirutil.py", line 39, in fast_relpath
    raise ValueError('{} is not a directory containing {}'.format(start, path))

Exception message: src/python/pants/backend/graph_info is not a directory containing src/rust/engine/src/lib.rs
```

where:

```
>>> os.path.relpath('src/rust/engine/src/lib.rs', 'src/python/pants/backend/graph_info')
'../../../../rust/engine/src/lib.rs'
```

vs

```>>> fast_relpath('src/rust/engine/src/lib.rs', 'src/python/pants/backend/graph_info')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/private/var/folders/bb/msvj_4_s741g93nzgrj19twc0000gn/T/tmpNb4Pj9/.deps/pantsbuild.pants-1.3.0-py2-none-any.whl/pants/util/dirutil.py", line 39, in fast_relpath
    raise ValueError('{} is not a directory containing {}'.format(start, path))
ValueError: src/python/pants/backend/graph_info is not a directory containing src/rust/engine/src/lib.rs
>>>
```

this was why I was alluding to an exact, faster `os.path.relpath` replica in #5570. but I think this is sufficient, unless it's possible that `EagerFilesetWithSpec._files` can contain the `../../` paths (which I'm assuming is not the case)?

Reply:
seems like falling back to `os.path.relpath` in this case (vs `return False`) is still a net perf win - happy to do that if you think it's more correct.

Reply:
err, actually not - scratch that last comment.

Reply:
Ooh, I see what is going on here. Yea, so not matching is expected.

Looks fine with the catch. The only potential improvement would be to have a `return_none=True` mode for `fast_relpath`, so that we don't hide other types of `ValueError` here. But also fine with just landing it.

Reply:
...oh. I already did that: https://github.com/pantsbuild/pants/blob/1889b938596746110696f9ff58c1985f76635b6a/src/python/pants/util/dirutil.py#L43

Reply:
ah, great idea - fixed.